samcli/lib/list/testable_resources/testable_resources_producer.py

```diff
@@ -0,0 +1,367 @@
+        return api_list
+
+    def produce(self):
+        """
+        The producer function for the testable resources command
+        """
+        sam_template = get_template_data(self.template_file)
+
+        translated_dict = self.get_translated_dict(template_file_dict=sam_template)
+        stacks, _ = SamLocalStackProvider.get_stacks(template_file="", template_dictionary=translated_dict)
+        validate_stack(stacks)
+        seen_testable_resources = set()
+        testable_resources_list = []
+        testable_resource_types = {"AWS::Lambda::Function", "AWS::ApiGateway::RestApi", "AWS::ApiGatewayV2::Api"}
+        if self.stack_name:
+            response = self.get_resources_info()
+            response_domain_dict = get_response_domain_dict(response)
+            custom_domain_substitute_dict = get_custom_domain_substitute_list(response, stacks, response_domain_dict)
+
+            for deployed_resource in response["StackResources"]:
+                if deployed_resource["ResourceType"] in testable_resource_types:
+                    endpoint_function_url = "-"
+                    paths_and_methods = "-"
+                    if deployed_resource["ResourceType"] == "AWS::Lambda::Function":
+                        endpoint_function_url = self.get_function_url(deployed_resource["PhysicalResourceId"])
+
+                    elif deployed_resource["ResourceType"] in ("AWS::ApiGateway::RestApi", "AWS::ApiGatewayV2::Api"):
+                        stages = self.get_stage_list(
+                            deployed_resource["PhysicalResourceId"],
+                            get_api_type_enum(deployed_resource["ResourceType"]),
+                        )
+                        if deployed_resource["LogicalResourceId"] in custom_domain_substitute_dict:
+                            endpoint_function_url = custom_domain_substitute_dict[
+                                deployed_resource["LogicalResourceId"]
+                            ]
+                        else:
+                            endpoint_function_url = self.build_api_gw_endpoints(
+                                deployed_resource["PhysicalResourceId"], stages
+                            )
+                        paths_and_methods = get_methods_and_paths(deployed_resource["LogicalResourceId"], stacks[0])
+
+                    testable_resource_data = TestableResDef(
+                        LogicalResourceId=deployed_resource["LogicalResourceId"],
+                        PhysicalResourceId=deployed_resource["PhysicalResourceId"],
+                        CloudEndpointOrFURL=endpoint_function_url,
+                        Methods=paths_and_methods,
+                    )
+                    testable_resources_list.append(dataclasses.asdict(testable_resource_data))
+                    seen_testable_resources.add(deployed_resource["LogicalResourceId"])
+            for local_resource in stacks[0].resources:
```

Comment:
stacks[0] sounds like a place where we could use some defensive programming.

Reply:
Is the validate_stack function adequate for checking this?